KAFKA-16082: Avoid resuming future replica if current replica is in the same directory

gaurav-narula opened the pull request:
KAFKA-16082 describes failure scenarios around replica directory being altered and a log directory/broker crashing.

It is observed that for scenario (3), i.e. a broker crashes while it waits for the future replica to catch up for the second time and the `dir1` is unavailable when the broker is restarted, the broker tries to create the partition in `dir2` according to the metadata in the controller. However, ReplicaManager also tries to resume the stale future replica which was abandoned when the broker crashed. This results in the renaming of the future replica to fail eventually because the directory for the topic partition already exists in `dir2` and the broker then marks `dir2` as offline.

This PR attempts to fix this behaviour by ignoring any future replicas which are in the same directory as where the log exists. It further marks the stale future replica for deletion.

Edit: Instead of removing the abandoned future replica, I've adopted Proven's suggestion to rename the future replica instead and avoid the creation of a new replica in `dir2`.

gaurav-narula committed:
KAFKA-16082: Avoid resuming future replica if current replica is in the
same directory

KAFKA-16082 describes failure scenarios around replica directory being
altered and a log directory/broker crashing.

It is observed that for scenario (3), i.e. a broker crashes while it
waits for the future replica to catch up for the second time and the
`dir1` is unavailable when the broker is restarted, the
broker tries to create the partition in `dir2` according to the metadata
in the controller. However, ReplicaManager also tries to resume the
stale future replica which was abandoned when the broker crashed. This
results in the renaming of the future replica to fail eventually because
the directory for the topic partition already exists in `dir2` and the
broker then marks `dir2` as offline.

This PR attempts to fix this behaviour by ignoring any future replicas
which are in the same directory as where the log exists. It further
marks the stale future replica for deletion.

gaurav-narula commented:
CC: @OmniaGM @pprovenzano

OmniaGM reviewed (approved):
This looks very straight forward assuming all tests are passing. LGTM

gaurav-narula committed:
KAFKA-16082: try renaming abandoned future replica

Instead of creating a new replica and deleting the abandoned future
replica, we try to rename the abandoned future replica instead.

This optimisation should help in faster catching up since we won't
be records from the beginning.

OmniaGM commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -1156,6 +1156,32 @@ class LogManager(logDirs: Seq[File],
+        assignedDirId <- partitionAssignedDirectoryId
+        if futureDirId == assignedDirId
+      } {
+        val sourceLog = futureLog
```
`sourceLog` can be moved into the `for` block.

OmniaGM commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -2773,6 +2776,12 @@ class ReplicaManager(val config: KafkaConfig,
         Some(partition, false)
 
       case HostedPartition.None =>
+        var isNew = true
```
`isNew` can be a val if you changed the code to the following
```
val isNew = if (isLocalFollower) {
           logManager.maybeRecoverAbandonedFutureLog(tp, partitionAssignedDirectoryId, () => {
             isNew = false
           })
         } else true
```

gaurav-narula replied:
I don't think that would work because `isNew` is mutated inside the callback which may be executed conditionally.

Edit: Perhaps I can have `maybeRecoverAbandonedFutureLog` return a boolean instead

OmniaGM commented on core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:
```diff
@@ -4831,15 +4831,15 @@ class ReplicaManagerTest {
       val foo0 = new TopicPartition("foo", 0)
       val emptyDelta = new TopicsDelta(TopicsImage.EMPTY)
-      val (fooPart, fooNew) = replicaManager.getOrCreatePartition(foo0, emptyDelta, FOO_UUID).get
+      val (fooPart, fooNew) = replicaManager.getOrCreatePartition(foo0, emptyDelta, FOO_UUID, false, None).get
```
Any reason why we are not setting `partitionAssignedDirectoryId` to None as a default in `replicaManager.getOrCreatePartition` signature?

OmniaGM commented on core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:
```diff
@@ -5592,6 +5592,63 @@ class ReplicaManagerTest {
+    val mockLogMgr = replicaManager.logManager
+    when(
+      mockLogMgr.getLog(topicPartition, isFuture = true)
+    ).thenReturn {
```
`{` is used to hold a block but in this case it is a simple value. So maybe replace all `.thenReturn {Some(mockFutureLog)}` to `.thenReturn(Some(mockFutureLog))`

OmniaGM commented on core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:
```diff
@@ -5592,6 +5592,63 @@ class ReplicaManagerTest {
+      Some(mockFutureLog)
+    }
+    when(mockLogMgr.getLog(topicPartition)).thenReturn {
+      None
```
Same `.thenReturn {None}` to `.thenReturn(None)`

OmniaGM commented on core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:
```diff
@@ -5592,6 +5592,63 @@ class ReplicaManagerTest {
+    when(
+      mockLogMgr.directoryId(mockFutureLog.parentDir)
+    ).thenReturn {
+      Some(directoryId1)
```
same as above

OmniaGM reviewed (commented):
left few scala nits

OmniaGM reviewed (approved):
LGTM, @pprovenzano @rondagostino can you please have a look?

pprovenzano reviewed (approved):
LGTM

OmniaGM commented:
Hi, @rondagostino, @showuon and @cmccabe can one of you have a look into this pr please and merge it if you are happy with the changes?

soarez commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -2859,10 +2867,10 @@ class ReplicaManager(val config: KafkaConfig,
     localLeaders.forKeyValue { (tp, info) =>
-      getOrCreatePartition(tp, delta, info.topicId).foreach { case (partition, isNew) =>
+      val partitionAssignedDirectoryId = directoryIds.find(_._1.topicPartition() == tp).map(_._2)
+      getOrCreatePartition(tp, delta, info.topicId, isLocalFollower = false, partitionAssignedDirectoryId).foreach { case (partition, isNew) =>
```
> `isLocalFollower = false`

I don't think this is ok. Setting this to false disables the fix when the broker is the leader. Which can happen in RF=1 or with RF=N when the broker comes back in the ISR (e.g. because there were no new messages, and the other replicas are all shutdown).

I had a conversation with @gaurav-narula about this, and the plan is to apply this fix earlier when the logs are being loaded.

gaurav-narula committed:
Perform recovery immediately after log load

soarez commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -2790,7 +2791,6 @@ class ReplicaManager(val config: KafkaConfig,
           //   is unavailable. This is required to ensure that we include the partition's
           //   high watermark in the checkpoint file (see KAFKA-1647).
           val state = info.partition.toLeaderAndIsrPartitionState(tp, isNew)
-          val partitionAssignedDirectoryId = directoryIds.find(_._1.topicPartition() == tp).map(_._2)
```
It seems the changes in this file are no longer necessary?

soarez commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -1173,6 +1173,35 @@ class LogManager(logDirs: Seq[File],
+        throw new RuntimeException(s"The log dir $log does not have a topic ID, " +
+          "which is not allowed when running in KRaft mode.")
+      }
+      val partitionId = log.topicPartition.partition()
+      Option(newTopicsImage.getPartition(topicId, partitionId))
+        .filter(pr => directoryId(log.parentDir).contains(pr.directory(brokerId)))
+        .map(_ => log)
```
1. Do we have the guarantee that the topic will be in the new topics image? Can't the topic be in later metadata delta if a compaction hasn't yet occurred?
2. Shouldn't we check if the main replica is on one of the other directories? If the main replica is on an offline dir (as in the scenario described in the issue) the broker will refuse to start once it is restarted with that directory online, as it will see the two main replicas. If the log directory for the main replica is online, we should be able to detect that here.

showuon replied:
For (1), I think we'll catch up until the high watermark before publish the image. So it should be safe.
For (2), good suggestion.

gaurav-narula replied:
Thanks for the feedback.

For (2), we've couple of options. We can either:

(a) ignore the future replica (say in dir2) if the main replica exists in an online log dir (say dir1) or,
(b) promote the future replica (in dir2)  and remove the main replica (in dir1).

(a) would result in ReplicaManager spawning a replicaAlterLogDir thread for the future replica and correcting the assignment to dir1, only for it to be changed back again to dir2 when the replicaAlterLogDir thread finishes its job. Refer https://github.com/apache/kafka/blob/acecd370cc3b25f12926e7a4664a2648f08c6c9a/core/src/main/scala/kafka/server/ReplicaManager.scala#L2734 and https://github.com/apache/kafka/blob/acecd370cc3b25f12926e7a4664a2648f08c6c9a/core/src/main/scala/kafka/server/ReplicaManager.scala#L2745

Since in these scenarios, the future replica is almost caught up with the main replica, I'm leaning towards option (b) to avoid more reassignments. Please let me know if you feel otherwise.

soarez replied:
If we didn't know if the future log was caught up or not, then I'd prefer (a), but at this point I can't conceive of a different scenario – other than a failure during replica promotion – that would cause the future log to be in the directory assigned in the metadata. So I agree that the two logs likely will be either caught up or very close. So I agree it makes more sense to do (b) - promote the future log and delete the main one.

We can still run into trouble if the directory with the main replica is offline. At some point that will cause a crash if the directory ever comes back online. But there's nothing we can do about that here. Maybe future work could improve how the broker handles loading conflicting logs.

gaurav-narula replied:
Addressed in 430546fa5a78118497ada6373607f6a25c78a8e8. Please take a look.

showuon replied:
I agree option (b) is better. But for this:
> We can still run into trouble if the directory with the main replica is offline. At some point that will cause a crash if the directory ever comes back online. But there's nothing we can do about that here. Maybe future work could improve how the broker handles loading conflicting logs.

I believe this PR: https://github.com/apache/kafka/pull/15335 should already fix this issue. @OmniaGM , could you help confirm it? If not, I think we should not promote the future log when main replica is offline to cause potential future issue.

soarez reviewed (commented):
Thanks for the changes. I still think there are some issues.

showuon commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -2745,10 +2745,10 @@ class ReplicaManager(val config: KafkaConfig,
+      val partitionAssignedDirectoryId = directoryIds.find(_._1.topicPartition() == tp).map(_._2)
       getOrCreatePartition(tp, delta, info.topicId).foreach { case (partition, isNew) =>
         try {
           val state = info.partition.toLeaderAndIsrPartitionState(tp, isNew)
```
Same here, unnecessary change.

showuon reviewed (commented):
Thanks for the PR. Left some comments.

gaurav-narula committed:
Delete older replica if found in online log dir

gaurav-narula committed:
Undo unneccesary change

soarez commented on core/src/test/java/kafka/testkit/BrokerNode.java:
```diff
@@ -81,8 +82,7 @@ public BrokerNode build(
                         singletonList(String.format("combined_%d", id));
                 } else {
-                    logDataDirectories = Collections.
-                        singletonList(String.format("broker_%d_data0", id));
+                    logDataDirectories = Collections.unmodifiableList(Arrays.asList(String.format("broker_%d_data0", id), String.format("broker_%d_data1", id)));
```
I'm wondering if this is going to break test cases that don't support JBOD, if so there should be some new failing tests.

chia7712 replied:
Personally, I'd like to change setNumBrokerNodes(int) to setBrokerNodes(int, int). The second argument is used to define the number of data folders. After all, not all tests require 2+ folders. Maybe we can address that in another PR.

chia7712 replied:
I have created https://issues.apache.org/jira/browse/KAFKA-16559

gaurav-narula replied:
Rebased to use the new method in the builder

soarez commented on core/src/test/scala/integration/kafka/server/KRaftClusterTest.scala:
```diff
@@ -1437,7 +1438,9 @@ class KRaftClusterTest {
-        // Modify foo-0 so that it refers to a future replica
+        // Modify foo-0 so that it refers to a future replica.
+        // This has the same effect as the main replica being in an offline
+        // log dir and the broker crashing just at the time of promotion
```
I think I know what you mean, but I don't know if this is easy to understand in its current phrasing. Here's a suggestion:

This is equivalent to a failure during the promotion of the future replica and a restart with directory for the main replica being offline.

soarez commented on core/src/test/scala/integration/kafka/server/KRaftClusterTest.scala:
```diff
@@ -1456,6 +1459,76 @@ class KRaftClusterTest {
+
+        val log = broker0.logManager.getLog(foo0).get
+
+        // Copy foo-0 to another log dir
+        val parentDir = log.parentDir
+        var targetParentDir = parentDir.substring(0, parentDir.length - 1)
+        if (parentDir.endsWith("0")) {
+          targetParentDir += "1"
+        } else {
+          targetParentDir += "0"
+        }
```
This assumes your change to `BrokerNode.build()` and its specific naming pattern. We'd get a more robust test if instead you access the directory list on the broker, filter out `parentDir`, and choose one of the remaining log dirs.

soarez commented on core/src/test/scala/integration/kafka/server/KRaftClusterTest.scala:
```diff
@@ -1456,6 +1459,76 @@ class KRaftClusterTest {
+
+        val log = broker0.logManager.getLog(foo0).get
+
+        // Copy foo-0 to another log dir
```
It could be helpful for future eyes on this, to leave a comment explaining why we're manually copying the log dir in the test.

gaurav-narula committed:
Reword comments in test

soarez reviewed (approved):
Thanks for the changes. The test failures seem to be unrelated. 
LGTM

soarez commented:
@showuon PTAL

showuon commented:
Will check it again within this week. Thanks.

showuon commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -1176,6 +1176,42 @@ class LogManager(logDirs: Seq[File],
+  def recoverAbandonedFutureLogs(brokerId: Int, newTopicsImage: TopicsImage): Unit = {
+    val abandonedFutureLogs = findAbandonedFutureLogs(brokerId, newTopicsImage)
+    abandonedFutureLogs.foreach { case (futureLog, currentLog) =>
+      val tp = futureLog.topicPartition
+
+      futureLog.renameDir(UnifiedLog.logDirName(tp), shouldReinitialize = true)
+      futureLog.removeLogMetrics()
+      futureLogs.remove(tp)
+
+      currentLog.foreach { log =>
+        log.removeLogMetrics()
+        log.renameDir(UnifiedLog.logDeleteDirName(tp), shouldReinitialize = false)
+        addLogToBeDeleted(log)
+        info(s"Old log for partition ${tp} is renamed to ${log.dir.getAbsolutePath} and is scheduled for deletion")
+      }
+
+      currentLogs.put(tp, futureLog)
+      futureLog.newMetrics()
+
+      info(s"Successfully renamed abandoned future log for $tp")
```
I can see we basically do the similar things as what `replaceCurrentWithFutureLog` did, why should we duplicate the codes here again? Could we not be able to call `replaceCurrentWithFutureLog` with additional parameters to indicate we want to skip some actions?

Like what we did now, we didn't close the `currentLog`, which should have potential resource leak. Using the `replaceCurrentWithFutureLog` could avoid this issue happening. WDYT?

chia7712 replied:
it seems to me that we need to call `closeHandlers` at least if we don't do refactor in this PR.

gaurav-narula replied:
That's a great suggestion! I'm going to try refactoring this to use `replaceCurrentWithFutureLog`

gaurav-narula replied:
I've refactored LogManager to use `replaceCurrentWithFutureLog` in [b87a21f](https://github.com/apache/kafka/pull/15136/commits/b87a21f3bd0eea6e4083a2d14b41053361f7b40a). PTAL

showuon reviewed (commented):
@gaurav-narula , thanks for the PR. Left one comment.

chia7712 commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -1176,6 +1176,42 @@ class LogManager(logDirs: Seq[File],
+        log.removeLogMetrics()
+        log.renameDir(UnifiedLog.logDeleteDirName(tp), shouldReinitialize = false)
+        addLogToBeDeleted(log)
+        info(s"Old log for partition ${tp} is renamed to ${log.dir.getAbsolutePath} and is scheduled for deletion")
```
`${tp}` -> `$tp`

gaurav-narula committed:
LogManager: refactor to use replaceCurrentWithFutureLog

chia7712 commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -1189,50 +1216,61 @@ class LogManager(logDirs: Seq[File],
+    futureLogs.remove(topicPartition)
+    currentLogs.put(topicPartition, destLog)
+    if (cleaner != null) {
+      cleaner.alterCheckpointDir(topicPartition, sourceLog.map(_.parentDirFile), destLog.parentDirFile)
```
It seems `cleaner.alterCheckpointDir` will do nothing if `sourceLog` is empty. Maybe we can revert those changes and run `alterCheckpointDir` only if `sourceLog` is defined.

gaurav-narula replied:
Addressed in [d02fc0b](https://github.com/apache/kafka/pull/15136/commits/d02fc0b9abeacf95096d4275232ea0cf829d836a)

chia7712 reviewed (commented):
@gaurav-narula thanks for refactoring code. one major comment left. please take a look.

gaurav-narula committed:
Invoke alterCheckpointDir if sourceDir is provided

chia7712 commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -1189,50 +1216,63 @@ class LogManager(logDirs: Seq[File],
-        resumeCleaning(topicPartition)
+  def replaceCurrentWithFutureLog(sourceLog: Option[UnifiedLog], destLog: UnifiedLog, updateHighWatermark: Boolean = false): Unit = {
+    val topicPartition = destLog.topicPartition
+    info(s"Attempting to replace current log $sourceLog with $destLog for $topicPartition")
```
`sourceLog` could be empty now, so maybe we need to revise the log message

gaurav-narula replied:
Addressed this and other comments in [31412a9](https://github.com/apache/kafka/pull/15136/commits/31412a9485b63731a49b5d44d6dc6fbeaf52dd0f)

chia7712 commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -1178,6 +1178,33 @@ class LogManager(logDirs: Seq[File],
+    val abandonedFutureLogs = findAbandonedFutureLogs(brokerId, newTopicsImage)
+    abandonedFutureLogs.foreach { case (futureLog, currentLog) =>
+      val tp = futureLog.topicPartition
+      if (cleaner != null) {
```
we can replace this `abortAndPauseCleaning(tp)`

chia7712 replied:
I guess `cleaner.abortAndPauseCleaning` is added because we call `resumeCleaning` later, and it will cause error if we don't call `abortAndPauseCleaning` here?

gaurav-narula replied:
> I guess cleaner.abortAndPauseCleaning is added because we call resumeCleaning later, and it will cause error if we don't call abortAndPauseCleaning here?

That's correct if the cleaning operation hasn't started. The cleaning operation is scheduled on a separate thread so we cannot be sure if `inProgress` map in `LogCleanerManager` has a key for the given topicPartition at the time we iterate over these logs.

chia7712 replied:
Maybe we need to add comments for `abortAndPauseCleaning`. The `recoverAbandonedFutureLogs` is called in starting and calling `abortAndPauseCleaning` here is a bit weird to me

gaurav-narula replied:
Added a comment in [3d78e55](https://github.com/apache/kafka/pull/15136/commits/3d78e5567fd16aa77f0b9f8c510ad26f0e6443c6)

chia7712 commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -1178,6 +1178,33 @@ class LogManager(logDirs: Seq[File],
+
+      replaceCurrentWithFutureLog(currentLog, futureLog)
+
+      info(s"Successfully renamed abandoned future log for $tp")
```
line#1189 will print `info(s"The current replica is successfully replaced with the future replica for $topicPartition")` and that is not much related to fact, right?

chia7712 reviewed (commented):
@gaurav-narula thanks for updated PR. please take a look at following comments.

gaurav-narula committed:
Improve logging

chia7712 commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -1189,49 +1217,62 @@ class LogManager(logDirs: Seq[File],
+    }
+  }
+
+  def replaceCurrentWithFutureLog(sourceLog: Option[UnifiedLog], destLog: UnifiedLog, updateHighWatermark: Boolean = false): Unit = {
```
it seems we don't need `updateHighWatermark`, since we call `updateHighWatermark` only if `sourceLog` is defined.

gaurav-narula replied:
I think that's not necessarily true. `findAbandonedFutureLogs` may find a `sourceLog` if the log directory is online but I don't think it's safe to update the high watermark even then because sourceLog may be ahead of futureLog.

chia7712 replied:
> I think that's not necessarily true. findAbandonedFutureLogs may find a sourceLog if the log directory is online but I don't think it's safe to update the high watermark even then because sourceLog may be ahead of futureLog.

you are right!

chia7712 reviewed (commented):
@gaurav-narula please take a look at two minor comments. thanks!

gaurav-narula committed:
Comment the usage of abortAndPauseCleaning

chia7712 reviewed (approved):
LGTM

chia7712 commented:
the failed `testParseAndValidateAddressesWithReverseLookup` is traced by #15758. will merge it
